zson/formatter.go

```diff
@@ -117,26 +117,46 @@ func (f *Formatter) Format(zv *zed.Value) (string, error) {
 }
 
 func (f *Formatter) hasName(typ zed.Type) bool {
-	ok := f.typedefs.exists(typ)
-	if !ok && f.persist != nil {
-		ok = f.permanent.exists(typ)
+	named, ok := typ.(*zed.TypeNamed)
+	if !ok {
+		return false
+	}
+	if _, ok := f.typedefs[named.Name]; ok {
+		return true
 	}
-	return ok
+	if f.persist != nil {
```

Comment:
Nit: Test f.permanent since that's the field that this guards.
```suggestion
	if f.permanent != nil {
```